codesearch/index/index.go

```diff
@@ -31,6 +31,12 @@ const batchFlushSizeBytes = 1_000_000_000 // flush batch every 1G
 
 type postingLists map[string]posting.List
 
+// Writer is not thread-safe. A single instance should not be used concurrently.
+// Multiple instances can be used concurrently without crashing, however CRUD operations are
+// not atomic, so index corruption can occur if multiple writers are used to modify the same
```

Comment:
nit: I think batch writes are atomic (https://pkg.go.dev/github.com/cockroachdb/pebble#hdr-Atomic_commit-Batch) so as long as a single document doesn't span batches I think we're OK here? ( I think this is the case but I could be mis-reading the code)

Reply:
It's more that doing an "UpdateDocument" will read, then write. So if you have 2 indexing jobs trying to replace the same document at the same time:

- P1: read old doc id (100)
- P2: read old doc id (100)
- P1: delete old doc id (100)
- P2: delete old doc id (100)
- P1: add new doc (101)
- P2: add new doc (102)

So you end up with 2 copies of the doc at the end.

Reply:
ahhhh got it. Yeah ok, agree we gotta do something better to fix that